receiver/dockerstatsreceiver/docker.go

```diff
@@ -0,0 +1,247 @@
+	containers           map[string]DockerContainer
+	excludedImageMatcher *StringMatcher
+	logger               *zap.Logger
+}
+
+func NewDockerClient(config *Config, logger *zap.Logger) (*DockerClient, error) {
+	client, err := docker.NewClientWithOpts(
+		docker.WithHost(config.Endpoint),
+		docker.WithVersion(dockerAPIVersion),
+		docker.WithHTTPHeaders(map[string]string{"User-Agent": userAgent}),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create docker client: %w", err)
+	}
+
+	excludedImageMatcher, err := NewStringMatcher(config.ExcludedImages)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine docker client excluded images: %w", err)
+	}
+
+	dc := &DockerClient{
+		client:               client,
+		config:               config,
+		logger:               logger,
+		containers:           make(map[string]DockerContainer),
+		excludedImageMatcher: excludedImageMatcher,
+	}
+
+	return dc, nil
+}
+
+// Provides a slice of DockerContainers to use for individual FetchContainerStats calls.
+func (dc *DockerClient) Containers() []DockerContainer {
+	containers := make([]DockerContainer, 0, len(dc.containers))
+	for _, container := range dc.containers {
+		containers = append(containers, container)
+	}
+	return containers
+}
+
+// LoadContainerList will load the initial running container maps for
+// inspection and establishing which containers warrant stat gathering calls
+// by the receiver.
+func (dc *DockerClient) LoadContainerList(ctx context.Context) error {
+	// Build initial container maps before starting loop
+	filters := dfilters.NewArgs()
+	filters.Add("status", "running")
+	options := dtypes.ContainerListOptions{
+		Filters: filters,
+	}
+
+	listCtx, cancel := context.WithTimeout(ctx, dc.config.Timeout)
+	containerList, err := dc.client.ContainerList(listCtx, options)
+	defer cancel()
+	if err != nil {
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+	for _, c := range containerList {
+		wg.Add(1)
+		go func(container dtypes.Container) {
+			if !dc.shouldBeExcluded(container.Image) {
+				if cnt, ok := dc.inspectedContainerIsOfInterest(ctx, container.ID); ok {
+					dc.persistContainer(cnt)
+				}
+			} else {
+				dc.logger.Debug(
+					"Not monitoring container per ExcludedImages",
+					zap.String("image", container.Image),
+					zap.String("id", container.ID),
+				)
+			}
+			wg.Done()
+		}(c)
+	}
+	wg.Wait()
+	return nil
+}
+
+// FetchContainerStatsAndConvertToMetrics will query the desired container stats and send
+// converted metrics to the results channel, since this is intended to be run in a goroutine.
+func (dc *DockerClient) FetchContainerStatsAndConvertToMetrics(
+	ctx context.Context,
+	container DockerContainer,
+	results chan<- Result,
+) {
+	dc.logger.Debug("Fetching container stats.", zap.String("id", container.ID))
+	statsCtx, cancel := context.WithTimeout(ctx, dc.config.Timeout)
+	containerStats, err := dc.client.ContainerStats(statsCtx, container.ID, false)
+	defer cancel()
```

Comment:
Is this `defer cancel()` needed?

Reply:
According to the docs it should be called to free resources: https://golang.org/pkg/context/#WithTimeout